### PR TITLE
Next version of Caporal

### PR DESCRIPTION
**Pull request by mattallty**

### CODE REVIEW

File: src/autocomplete.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+   * @param {Argument|Option} arg_or_opt argument or option to complete
+   */
+  getCompletion(arg_or_opt) {
+    return this._completions[arg_or_opt.name()]
+  }
+
+  _findCommand(cmdStr) {
+    let cmd
+    const commandArr = []
+    const args = cmdStr.split(" ").filter(s => s != "")
```
Comment (mattallty) [inline src/autocomplete.js:43]: Codacy Issue found: [Replace `s` with `(s)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

**Comment by mattallty** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 26
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187&bid=17918968)

File: src/autocomplete.js [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+      return Promise.resolve([])
+    }
+
+    // Choices
+    if (arg.getChoices().length) {
+      return Promise.resolve(
+        arg
+          .getChoices()
+          .map(choice => choice + ":Value for argument " + arg.synopsis())
+          .filter(choice => !lastPartial || choice.startsWith(lastPartial))
```
Comment (mattallty) [inline src/autocomplete.js:139]: Codacy Issue found: [Replace `(choice·=>·!lastPartial·||·choice.startsWith(lastPartial))` with `((choice)·=>·!lastPartial·||·choice.startsWith(lastPartial)),`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    const lastPartIsOption = data.lastPartial.startsWith("-")
+    const lastPartIsKnownOption = this._lastPartialIsKnownOption(
+      currCommand,
+      lastPartIsOption,
+      lastPartial
+    )
+    const currOption = this._getCurrentOption(
+      currCommand,
+      lastPartIsKnownOption,
+      lastPartial
```
Comment (mattallty) [inline src/autocomplete.js:261]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: babel.config.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      {
+        "targets": {
+          "node": 10
+        }
+      }
+    ],
+    "@babel/preset-typescript"
+  ],
+  "plugins": ["@babel/plugin-proposal-class-properties"],
+  "ignore": ["**/__tests__", "**/*.spec.{t,j}s"]
```
Comment (mattallty) [inline babel.config.json:14]: Codacy Issue found: [Format Error: expected "\n  " ](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    return Promise.all([
+      possCommands,
+      possArgValues,
+      possOptNames,
+      possOptValues
+    ])
+      .then(function (results) {
+        const completions = [].concat
+          .apply([], results)
+          .filter(e => typeof e != "undefined")
```
Comment (mattallty) [inline src/autocomplete.js:288]: Codacy Issue found: [Replace `e` with `(e)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  _findCommand(cmdStr) {
+    let cmd
+    const commandArr = []
+    const args = cmdStr.split(" ").filter(s => s != "")
+    const commands = this._program.getCommands()
+
+    while (!cmd && args.length) {
+      commandArr.push(args.shift())
+      const cmdStr = commandArr.join(" ")
+      cmd = commands.filter(c => {
```
Comment (mattallty) [inline src/autocomplete.js:49]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+      currCommand,
+      lastPartIsOption,
+      realArgsCount,
+      data.lastPartial
+    )
+    const possOptNames = this._getPossibleOptionNames(
+      currCommand,
+      optionsAlreadyUsed,
+      lastPartial,
+      lastPartIsOption
```
Comment (mattallty) [inline src/autocomplete.js:274]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        if (!cmdStr) {
+          return true
+        }
+        return (
+          (c.name().startsWith(cmdStr) ||
+            (c.getAlias() && c.getAlias().startsWith(cmdStr))) &&
+          (!currentCommand || !this._isSameCommand(c, currentCommand))
+        )
+      })
+      .filter(c => c.name() !== "") // do not take the default command
```
Comment (mattallty) [inline src/autocomplete.js:228]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+      .map(c => {
```
Comment (mattallty) [inline src/autocomplete.js:229]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: babel.config.json [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+      "@babel/preset-env",
+      {
+        "targets": {
+          "node": 10
+        }
+      }
+    ],
+    "@babel/preset-typescript"
+  ],
+  "plugins": ["@babel/plugin-proposal-class-properties"],
```
Comment (mattallty) [inline babel.config.json:13]: Codacy Issue found: [Format Error: expected "\n  " ](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+      }
+      return res.map(r => r + `:Value for ${type} ${argOrOpt.synopsis()}`)
+    })
+  }
+
+  _getPossibleOptionNames(
+    currentCommand,
+    optionsAlreadyUsed,
+    lastPartial,
+    lastPartialIsOption
```
Comment (mattallty) [inline src/autocomplete.js:168]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+              return o.getLongOrShortName() + ":" + o.description()
+            } else if (
+              !lastPartialIsOption &&
+              optionsAlreadyUsed.indexOf(o.getShortName()) === -1 &&
+              optionsAlreadyUsed.indexOf(o.getLongName()) === -1
+            ) {
+              return o.getLongOrShortName() + ":" + o.description()
+            }
+          })
+          .filter(o => typeof o != "undefined")
```
Comment (mattallty) [inline src/autocomplete.js:191]: Codacy Issue found: [Replace `(o` with `((o)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+      return this._hanldleCompletionHandler(completion, currentOption)
+    }
+
+    return Promise.resolve([])
+  }
+
+  _getPossibleCommands(currentCommand, cmdStr) {
+    const commands = this._program
+      .getCommands()
+      .filter(c => {
```
Comment (mattallty) [inline src/autocomplete.js:218]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        )
+    )
+  }
+
+  _getCurrentOption(currentCommand, lastPartialIsKnownOption, lastPartial) {
+    return lastPartialIsKnownOption
+      ? currentCommand
+          .options()
+          .filter(
+            o =>
```
Comment (mattallty) [inline src/autocomplete.js:111]: Codacy Issue found: [Replace `o·=>⏎··············lastPartial·===·o.getShortName()·||⏎··············lastPartial·===·o.getLongName()` with `(o)·=>·lastPartial·===·o.getShortName()·||·lastPartial·===·o.getLongName(),`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+      currCommand,
+      lastPartIsKnownOption,
+      lastPartial
+    )
+
+    const possArgValues = this._getPossibleArgumentValues(
+      currCommand,
+      lastPartIsOption,
+      realArgsCount,
+      data.lastPartial
```
Comment (mattallty) [inline src/autocomplete.js:268]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+      possOptValues
+    ])
+      .then(function (results) {
+        const completions = [].concat
+          .apply([], results)
+          .filter(e => typeof e != "undefined")
+        done(null, completions)
+        return completions
+      })
+      .catch(err => {
```
Comment (mattallty) [inline src/autocomplete.js:292]: Codacy Issue found: [Replace `err` with `(err)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+    if (!arg) {
+      return Promise.resolve([])
+    }
+
+    // Choices
+    if (arg.getChoices().length) {
+      return Promise.resolve(
+        arg
+          .getChoices()
+          .map(choice => choice + ":Value for argument " + arg.synopsis())
```
Comment (mattallty) [inline src/autocomplete.js:138]: Codacy Issue found: [Replace `choice` with `(choice)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+    )
+  }
+
+  _lastPartialIsKnownArg(currentCommand, lastPartialIsArg, lastPartial) {
+    return (
+      lastPartialIsArg &&
+      currentCommand &&
+      currentCommand
+        .args()
+        .some(
```
Comment (mattallty) [inline src/autocomplete.js:100]: Codacy Issue found: [Replace `⏎··········a·=>·a.getChoices().length·&&·a.getChoices().includes(lastPartial)⏎········` with `(a)·=>·a.getChoices().length·&&·a.getChoices().includes(lastPartial)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+const parseArgs = require("micromist")
+const Promise = require("../tests/node_modules/bluebird")
+const Option = require("./option")
+
+class Autocomplete {
+  constructor(program) {
+    this._program = program
+    this._tab = tabtab({
+      name: this._program.bin(),
+      cache: false
```
Comment (mattallty) [inline src/autocomplete.js:13]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  _hanldleCompletionHandler(handler, argOrOpt) {
+    const result = handler()
+    const type = argOrOpt instanceof Option ? "option" : "argument"
+    return Promise.resolve(result).then(res => {
+      if (!Array.isArray(res)) {
+        res = [res]
+      }
+      return res.map(r => r + `:Value for ${type} ${argOrOpt.synopsis()}`)
```
Comment (mattallty) [inline src/autocomplete.js:160]: Codacy Issue found: [Replace `(r` with `((r)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+  _getPossibleOptionNames(
+    currentCommand,
+    optionsAlreadyUsed,
+    lastPartial,
+    lastPartialIsOption
+  ) {
+    const optNames = currentCommand
+      ? currentCommand
+          .options()
+          .map(o => {
```
Comment (mattallty) [inline src/autocomplete.js:173]: Codacy Issue found: [Replace `o` with `(o)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+    const possOptNames = this._getPossibleOptionNames(
+      currCommand,
+      optionsAlreadyUsed,
+      lastPartial,
+      lastPartIsOption
+    )
+    const possOptValues = this._getPossibleOptionValues(currOption)
+    const possCommands = this._getPossibleCommands(currCommand, cmd)
+
+    return Promise.all([
```
Comment (mattallty) [inline src/autocomplete.js:279]: Codacy Issue found: [Replace `⏎······possCommands,⏎······possArgValues,⏎······possOptNames,⏎······possOptValues⏎····` with `possCommands,·possArgValues,·possOptNames,·possOptValues`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 23]
```diff
@@ -0,0 +1,299 @@
...
+        realArgsCmdFound = true
+      } else if (realArgsCmdFound && value.trim()) {
+        acc++
+      }
+      return acc
+    }, 0)
+  }
+
+  _getOptionsAlreadyUsed(args) {
+    return args.filter(o => o.startsWith("-") || o.startsWith("--"))
```
Comment (mattallty) [inline src/autocomplete.js:78]: Codacy Issue found: [Replace `(o` with `((o)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+    return cmd
+  }
+
+  _countArgs(currentCommand, args) {
+    let realArgsCmdFound = false
+    return args._.reduce((acc, value, index, arr) => {
+      const possibleCmd = arr.slice(0, index + 1).join(" ")
+      if (
+        currentCommand &&
+        (currentCommand.name() === possibleCmd ||
```
Comment (mattallty) [inline src/autocomplete.js:66]: Codacy Issue found: [Delete `⏎·········`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

**Comment by mattallty** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 26
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187&bid=17918968)

File: src/autocomplete.js (hunk 14 above)
Comment (mattallty) [inline src/autocomplete.js:111]: Codacy Issue found: [Replace `o·=>⏎··············lastPartial·===·o.getShortName()·||⏎··············lastPartial·===·o.getLongName()` with `(o)·=>·lastPartial·===·o.getShortName()·||·lastPartial·===·o.getLongName(),`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 22 above)
Comment (mattallty) [inline src/autocomplete.js:279]: Codacy Issue found: [Replace `⏎······possCommands,⏎······possArgValues,⏎······possOptNames,⏎······possOptValues⏎····` with `possCommands,·possArgValues,·possOptNames,·possOptValues`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+      return this._hanldleCompletionHandler(completion, arg)
+    }
+
+    return Promise.resolve([])
+  }
+
+  _hanldleCompletionHandler(handler, argOrOpt) {
+    const result = handler()
+    const type = argOrOpt instanceof Option ? "option" : "argument"
+    return Promise.resolve(result).then(res => {
```
Comment (mattallty) [inline src/autocomplete.js:156]: Codacy Issue found: [Replace `res` with `(res)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 16 above)
Comment (mattallty) [inline src/autocomplete.js:292]: Codacy Issue found: [Replace `err` with `(err)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 8 above)
Comment (mattallty) [inline src/autocomplete.js:228]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 13 above)
Comment (mattallty) [inline src/autocomplete.js:218]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 24 above)
Comment (mattallty) [inline src/autocomplete.js:66]: Codacy Issue found: [Delete `⏎·········`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 19 above)
Comment (mattallty) [inline src/autocomplete.js:13]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 18 above)
Comment (mattallty) [inline src/autocomplete.js:100]: Codacy Issue found: [Replace `⏎··········a·=>·a.getChoices().length·&&·a.getChoices().includes(lastPartial)⏎········` with `(a)·=>·a.getChoices().length·&&·a.getChoices().includes(lastPartial)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 2 above)
Comment (mattallty) [inline src/autocomplete.js:139]: Codacy Issue found: [Replace `(choice·=>·!lastPartial·||·choice.startsWith(lastPartial))` with `((choice)·=>·!lastPartial·||·choice.startsWith(lastPartial)),`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: babel.config.json (hunk 10 above)
Comment (mattallty) [inline babel.config.json:13]: Codacy Issue found: [Format Error: expected "\n  " ](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 11 above)
Comment (mattallty) [inline src/autocomplete.js:168]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 3 above)
Comment (mattallty) [inline src/autocomplete.js:261]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 9 above)
Comment (mattallty) [inline src/autocomplete.js:229]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+
+  _isSameCommand(command, command2) {
+    return command2.name() === command.name()
+  }
+
+  _complete(data, done) {
+    const currCommand = this._findCommand(data.args.slice(3).join(" "))
+    const args = parseArgs(
+      data.args.slice(1),
+      currCommand ? currCommand.parseArgsOpts : {}
```
Comment (mattallty) [inline src/autocomplete.js:244]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 15 above)
Comment (mattallty) [inline src/autocomplete.js:268]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 7 above)
Comment (mattallty) [inline src/autocomplete.js:274]: Codacy Issue found: [Insert `,`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: babel.config.json (hunk 4 above)
Comment (mattallty) [inline babel.config.json:14]: Codacy Issue found: [Format Error: expected "\n  " ](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 20 above)
Comment (mattallty) [inline src/autocomplete.js:160]: Codacy Issue found: [Replace `(r` with `((r)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 12 above)
Comment (mattallty) [inline src/autocomplete.js:191]: Codacy Issue found: [Replace `(o` with `((o)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 21 above)
Comment (mattallty) [inline src/autocomplete.js:173]: Codacy Issue found: [Replace `o` with `(o)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 6 above)
Comment (mattallty) [inline src/autocomplete.js:49]: Codacy Issue found: [Replace `c` with `(c)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 17 above)
Comment (mattallty) [inline src/autocomplete.js:138]: Codacy Issue found: [Replace `choice` with `(choice)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 1 above)
Comment (mattallty) [inline src/autocomplete.js:43]: Codacy Issue found: [Replace `s` with `(s)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: src/autocomplete.js (hunk 5 above)
Comment (mattallty) [inline src/autocomplete.js:288]: Codacy Issue found: [Replace `e` with `(e)`](https://app.codacy.com/manual/matthiasetienne/Caporal-js/pullRequest?prid=5467187)

File: docs/.vuepress/config.js [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+    // displayAllHeaders: true,
+    logo: "/assets/img/caporal.svg",
+    nav: [
+      { text: "Guide", link: "/guide/" },
+      { text: "API Reference", link: "/api/" },
+      { text: "Playground", link: "/playground" },
+    ],
+    // Assumes GitHub. Can also be a full GitLab url.
+    repo: "mattallty/Caporal.js",
+    repoLabel: "GithHub Project",
```
Comment (kinok) [inline docs/.vuepress/config.js:82]: typo here @mattallty

**Pull request closed without merging**